### ISSUE TITLE
How is Token-Level KL different from Sequence-Level KL?

### ISSUE THREAD

**Issue by mnoukhov** [huggingface/trl#2562]
### System Info

#2552 Introduced a "token-level KL" which seems to be exactly the same as "sequence-level KL". Since `kl_coef` is a scalar these two seem identical. Am I missing something? @qgallouedec 

https://github.com/huggingface/trl/blob/main/trl/trainer/rloo_trainer.py#L398-400

I personally can't figure out this differs from the original KL formulations (https://arxiv.org/pdf/1611.02796 and https://arxiv.org/abs/1907.00456)

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
                if args.token_level_kl:
                    # Token-level KL penalty: apply KL penalty per token
                    token_kl_penalty = -args.kl_coef * kl
                    non_score_reward = token_kl_penalty.sum(1)
                else:
                    # Sequence-level KL penalty: sum KL across tokens first
                    sequence_kl = kl.sum(1)
                    non_score_reward = -args.kl_coef * sequence_kl
```

### Expected behavior

These two are functionally the same since `args.kl_coef` is a scalar

### Checklist

- [X] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [X] I have included my system information
- [X] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any traceback provided is complete

**Comment by kashif** [comment]
@mnoukhov you are right! my bad... so I am thinking:
```py
                # Compute total reward with KL penalty
                if args.token_level_kl:
                    # Token-level KL penalty: apply KL penalty per token
                    kl_reward = -args.kl_coef * kl
                    # Apply reward at the last non-padded token position for each sequence
                    eos_indices = sequence_lengths.unsqueeze(1) - 1
                    last_reward = torch.zeros_like(kl).scatter_(dim=1, index=eos_indices, src=scores.unsqueeze(1))
                    non_score_reward = kl_reward.sum(1) + last_reward.sum(1)
                else:
                    # Sequence-level KL penalty: sum KL across tokens first
                    sequence_kl = kl.sum(1)
                    non_score_reward = -args.kl_coef * sequence_kl + scores
                rlhf_reward = non_score_reward
```

**Comment by mnoukhov** [comment]
@kashif this solution (and #2575) once again does not do anything functionally different.

All this code is doing is the same sum(KL) + score but in a roundabout way: creating a vector of 0 with score as the last value, and adding this to the vector of KL losses, then doing a sum. 

I do not know what the REINFORCE++ author meant by "token-level KL" but by all accounts, it is just the standard KL loss we have been using and should not be a separate option.

**Comment by kashif** [comment]
ah damn ok @mnoukhov  let me check.. i also added some tests to compare the results and will double check

**Comment by mnoukhov** [comment]
I just don't understand what the other formulation could be? Like, what other way is there to apply KL other than KL between the logprobs of the outputs under the two models and then a sum?
